sign/mayo/mode1/internal/matrix.go

```diff
@@ -0,0 +1,289 @@
+			}
+			for r := 0; r <= rBound; r++ {
+				tab := mulTable(m2[k*colsStride+c])
+				pos := r*cols + c
+				if isM1Triangular {
+					pos = r*(cols*2-r+1)/2 + (c - r)
+				}
+				vecMulAddPackedTab(P, m1[P*pos:], tab, acc[P*(r*cols2+k):])
+			}
+		}
+	}
+}
+
+// acc += (M1+M1^T)*M2
+// M1 of rows x rows is upper triangular; M2 is of rows x cols2
+// acc and M1 are multiple matrices, M2 is a single matrix
+func mulAddMUpperTriangularWithTransposeMatXMat(acc []uint64, m1 []uint64, m2 []uint8, rows int, cols2 int) {
+	m1pos := 0
+	for r := 0; r < rows; r++ {
+		for c := r; c < rows; c++ {
+			if c == r {
+				m1pos += 1
+				continue
+			}
+			for k := 0; k < cols2; k++ {
+				vecMulAddPacked(P, m1[P*m1pos:], m2[c*cols2+k], acc[P*(r*cols2+k):])
+				vecMulAddPacked(P, m1[P*m1pos:], m2[r*cols2+k], acc[P*(c*cols2+k):])
+			}
+			m1pos++
+		}
+	}
+}
+
+func mulAddMatVec(acc []byte, m []byte, v []byte, rows, cols int) {
+	for i := 0; i < rows; i++ {
+		for j := 0; j < cols; j++ {
+			acc[i] ^= byte(mulAddPacked(uint64(m[i*cols+j]), v[j]))
+		}
+	}
+}
+
+func upper(in []uint64, out []uint64, size int) {
+	pos := 0
+	for r := 0; r < size; r++ {
+		for c := r; c < size; c++ {
+			copy(out[P*pos:][:P], in[P*(r*size+c):][:P])
+			if r != c {
+				for p := 0; p < P; p++ {
+					out[P*pos+p] ^= in[P*(c*size+r)+p]
+				}
+			}
+			pos++
+		}
+	}
+}
+
+func variableTime(sps []uint64, p1 []uint64, p2 []uint64, p3 []uint64, s []uint8) {
+	var accumulator [K * N][P * 16]uint64
+
+	// compute P * S^t = [ P1  P2 ] * [S1] = [P1*S1 + P2*S2]
+	//                   [  0  P3 ]   [S2]   [        P3*S2]
+
+	// Note that S = S1||S2 is strided at N=V+O
+
+	// P1 * S1^t : VxV * V*K, where P1 is triangular
+	pos := 0
+	for r := 0; r < V; r++ {
+		for c := r; c < V; c++ {
+			for k := 0; k < K; k++ {
+				vecAddPacked(P, p1[P*pos:], accumulator[r*K+k][P*int(s[k*N+c]):])
+			}
+			pos++
+		}
+	}
+
+	// P2 * S2^t : V*O * O*K
+	pos = 0
+	for r := 0; r < V; r++ {
+		for c := 0; c < O; c++ {
+			for k := 0; k < K; k++ {
+				vecAddPacked(P, p2[P*pos:], accumulator[r*K+k][P*int(s[k*N+V+c]):])
+			}
+			pos++
+		}
+	}
+
+	// P3 * S2^t : O*O * O*K, where P3 is triangular
+	pos = 0
+	for r := 0; r < O; r++ {
+		for c := r; c < O; c++ {
+			for k := 0; k < K; k++ {
+				vecAddPacked(P, p3[P*pos:], accumulator[(r+V)*K+k][P*int(s[k*N+V+c]):])
+			}
+			pos++
+		}
+	}
+
+	for i := 0; i < K*N; i++ {
+		aggregate(P, accumulator[i], sps[P*i:])
+	}
+}
+
+func variableTime2(sps []uint64, s []uint8, pst []uint64) {
+	var accumulator [K * K][P * 16]uint64
+
+	// S * PST : KxN * N*K
+	for r := 0; r < K; r++ {
+		for c := 0; c < N; c++ {
+			for k := 0; k < K; k++ {
+				vecAddPacked(P, pst[P*(c*K+k):], accumulator[r*K+k][P*int(s[r*N+c]):])
+			}
+		}
+	}
+
+	for i := 0; i < K*K; i++ {
+		aggregate(P, accumulator[i], sps[P*i:])
+	}
+}
+
+// p is always P, but is still kept to be consistent with other functions
+//
+//nolint:unparam
+func vecAddPacked(p int, in []uint64, acc []uint64) {
```

Comment:
If `p` is always `P`, then just use `P` and not have the parameter.

Reply:
Here a design choice was made so that this is consistent with other similar functions which can be sometimes called with `p != P`.